File: lib/models/diagnosis_result.dart
/// Immutable value object representing a plant disease diagnosis.
///
/// Encapsulates the full output of a single inference run — the predicted
/// label, confidence score, and convenience properties derived from label
/// parsing (crop name, disease name, health status).
class DiagnosisResult {
  /// Raw label string from `labels.txt` (e.g. `"Tomato___Late_blight"`).
  final String rawLabel;

  /// Model confidence in the range `[0.0, 1.0]`.
  final double confidence;

  /// Detailed symptoms of the disease (if applicable).
  final String? symptoms;

  /// Underlying cause or pathogen of the disease (if applicable).
  final String? causes;

  /// Recommended management and treatment steps (if applicable).
  final String? management;

  const DiagnosisResult({
    required this.rawLabel,
    required this.confidence,
    this.symptoms,
    this.causes,
    this.management,
  });

  // ---------------------------------------------------------------------------
  // Derived properties
  // ---------------------------------------------------------------------------

  /// Human‑readable crop name extracted from the raw label.
  ///
  /// Splits on `"___"`, takes the first segment and replaces underscores with
  /// spaces.  e.g. `"Cherry_(including_sour)"` → `"Cherry (including sour)"`.
  String get cropName {
    final parts = rawLabel.split('___');
    return _humanize(parts.first);
  }

  /// Human‑readable disease name.
  ///
  /// Returns `"Healthy"` for labels whose second segment is `"healthy"`,
  /// otherwise the humanised disease string.
  String get diseaseName {
    final parts = rawLabel.split('___');
    if (parts.length < 2) return 'Unknown';
    final raw = parts[1];
    if (raw.toLowerCase() == 'healthy') return 'Healthy';
    return _humanize(raw);
  }

  /// Formatted display label combining crop and disease.
  String get displayLabel => '$cropName — $diseaseName';

  /// Whether this diagnosis indicates a healthy plant.
  bool get isHealthy => rawLabel.toLowerCase().contains('healthy');

  /// Confidence expressed as a percentage integer (0‑100).
  int get confidencePercent => (confidence * 100).round();

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  /// Converts an underscore‑delimited segment into a human‑readable string,
  /// preserving parenthetical expressions.
  static String _humanize(String segment) {
    return segment
        .replaceAll('_', ' ')
        .replaceAll('  ', ' ')
        .trim();
  }

  @override
  String toString() => 'DiagnosisResult($displayLabel, $confidencePercent%)';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is DiagnosisResult &&
          rawLabel == other.rawLabel &&
          confidence == other.confidence &&
          symptoms == other.symptoms &&
          causes == other.causes &&
          management == other.management;

  @override
  int get hashCode =>
      rawLabel.hashCode ^
      confidence.hashCode ^
      symptoms.hashCode ^
      causes.hashCode ^
      management.hashCode;

  Map<String, dynamic> toJson() => {
        'rawLabel': rawLabel,
        'confidence': confidence,
        'symptoms': symptoms,
        'causes': causes,
        'management': management,
      };

  factory DiagnosisResult.fromJson(Map<String, dynamic> json) {
    return DiagnosisResult(
      rawLabel: json['rawLabel'] as String,
      confidence: (json['confidence'] as num).toDouble(),
      symptoms: json['symptoms'] as String?,
      causes: json['causes'] as String?,
      management: json['management'] as String?,
    );
  }
}

File: landing/src/models/diagnosisResult.ts
import type { DiagnosisResultData, DiseaseInfo } from '../types/diagnosis';

/** Immutable, display-ready representation of one prediction. */
export class DiagnosisResult {
  readonly rawLabel: string;
  readonly confidence: number;
  readonly symptoms: string | null;
  readonly causes: string | null;
  readonly management: string | null;

  constructor(data: DiagnosisResultData) {
    this.rawLabel = data.rawLabel;
    this.confidence = data.confidence;
    this.symptoms = data.symptoms;
    this.causes = data.causes;
    this.management = data.management;
  }

  get cropName(): string {
    return humanize(this.rawLabel.split('___')[0] ?? '');
  }

  get diseaseName(): string {
    const parts = this.rawLabel.split('___');
    if (parts.length < 2) {
      return 'Unknown';
    }

    const disease = parts[1] ?? '';
    return disease.toLowerCase() === 'healthy' ? 'Healthy' : humanize(disease);
  }

  get displayLabel(): string {
    return `${this.cropName} — ${this.diseaseName}`;
  }

  get isHealthy(): boolean {
    return this.rawLabel.toLowerCase().includes('healthy');
  }

  get confidencePercent(): number {
    return Math.round(this.confidence * 100);
  }

  toJSON(): DiagnosisResultData {
    return {
      rawLabel: this.rawLabel,
      confidence: this.confidence,
      symptoms: this.symptoms,
      causes: this.causes,
      management: this.management,
    };
  }

  static fromJSON(value: unknown): DiagnosisResult {
    if (!isRecord(value) || typeof value.rawLabel !== 'string' || value.rawLabel.length === 0 || typeof value.confidence !== 'number' || !Number.isFinite(value.confidence)) {
      throw new TypeError('Invalid diagnosis result.');
    }

    return new DiagnosisResult({
      rawLabel: value.rawLabel,
      confidence: value.confidence,
      symptoms: optionalString(value.symptoms),
      causes: optionalString(value.causes),
      management: optionalString(value.management),
    });
  }

  static fromPrediction(rawLabel: string, confidence: number, info?: DiseaseInfo): DiagnosisResult {
    return new DiagnosisResult({
      rawLabel,
      confidence,
      symptoms: info?.symptoms ?? null,
      causes: info?.causes ?? null,
      management: info?.management ?? null,
    });
  }
}

export function humanize(segment: string): string {
  return segment.replaceAll('_', ' ').replace(/ {2,}/g, ' ').trim();
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function optionalString(value: unknown): string | null {
  return typeof value === 'string' && value.length > 0 ? value : null;
}
